### PR TITLE
Python: Improve error messages

### PR DESCRIPTION
**Pull request by KGB33**
Improves error messages when dagger is not installed or not running.

Also creates the catch all (`DaggerExecption`) mentioned in #3805.

Discord thread: https://discord.com/channels/707636530424053791/1041044357345718282/1041044783155654766

# Message examples (Tracebacks removed)
### Before
```
FileNotFoundError: [Errno 2] No such file or directory: 'docker'
```
### After
```
dagger.connectors.docker.ProvisionError: Command 'docker' not found.
```

### Before
```
dagger.connectors.docker.ProvisionError: Failed to copy engine session binary: None
```
### After 
```
dagger.connectors.docker.ProvisionError: Failed to copy engine session binary: docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
See 'docker run --help'.
```

### Before
```
ValueError: invalid literal for int() with base 10: ''
```
### After
```
dagger.exceptions.DaggerException: Error: failed to run container: docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
```

### CODE REVIEW

File: sdk/python/src/dagger/connectors/docker.py [hunk 1]
```diff
@@ -124,12 +127,21 @@ def start(self) -> None:
...
-        # read port number from first line of stdout
-        port = int(self._proc.stdout.readline())
+        try:
+            # read port number from first line of stdout
+            port = int(self._proc.stdout.readline())
+        except ValueError as e:
+            # Check if the subprocess exited with an error.
+            return_value = self._proc.poll()
+            if return_value != 0 or return_value is not None:
+                raise DaggerException(self._proc.stderr.readline()) from e
```
Comment (sipsma) [inline sdk/python/src/dagger/connectors/docker.py:141]: In the case where `self.cfg.log_output` is actually set, I'm not sure this will always work as expected (e.g. what if the user provides a `log_output` that has already been read from).

I was making a somewhat similar change to this in the Go SDK; there I ended up implementing it such that the stderr output is copied both to the `logOutput` and to a temporary buffer that's closed at the end of provisioning. I then use that temp buffer to include the output in the error message if something goes wrong. Code here: https://github.com/dagger/dagger/pull/3826/files#diff-998a9fff7d631c3029d76dc70642afdc4566b71900ca8dccc2837fb45bf027f4R75

I have no idea if it's straightforward to do that in Python too. If it is, might be a good change. If it isn't, then I'd personally be totally fine with just skipping the inclusion of these lines when `log_output` was explicitly set by the user. In that case, the user will just have to go look at that output to see what went wrong, which isn't ideal but still much better than the current state of things.

**Review by sipsma — COMMENTED** [review]
Left one comment and need Helder to review the python-aspects of the code but the general idea looks great to me, greatly appreciated!

File: sdk/python/src/dagger/connectors/docker.py (hunk 1 above)
Comment (KGB33) [reply sdk/python/src/dagger/connectors/docker.py]: You're right, when `self.cfg.log_output` was set it did not work as expected. 
I couldn't find a way to include the error message in both places. 

In the end the best solution I came up with was to include the error message from the subprocess if `self._proc.stderr` existed and was readable. If its not readable, include a hint instead (the full error message will still be in whatever `log_output' is set to).

## Traceback with readable `self._proc.stderr`
```
❯ python examples/client-simple/say.py Foo
Traceback (most recent call last):
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connectors/docker.py", line 136, in start
    port = int(self._proc.stdout.readline())
ValueError: invalid literal for int() with base 10: ''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/kgb33/Code/dagger/sdk/python/examples/client-simple/say.py", line 32, in <module>
    anyio.run(main, sys.argv[1:])
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_core/_eventloop.py", line 70, in run
    return asynclib.run(func, *args, **backend_options)
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 292, in run
    return native_run(wrapper(), debug=debug)
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 287, in wrapper
    return await func(*args)
  File "/home/kgb33/Code/dagger/sdk/python/examples/client-simple/say.py", line 11, in main
    async with dagger.Connection() as client:
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connection.py", line 39, in __aenter__
    return await self.connector.connect()
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connectors/docker.py", line 186, in connect
    await anyio.to_thread.run_sync(self.provision_sync)
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connectors/docker.py", line 202, in provision_sync
    self.engine.start()
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connectors/docker.py", line 142, in start
    raise DaggerException(
dagger.exceptions.DaggerException: Dagger engine failed to start. Error: failed to run container: docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
```
## Traceback with unreadable `self._proc.stderr` (w/ `Dagger.Config(log_output=sys.stderr)`)
```
❯ python examples/client-simple/say.py Foo
Error: failed to run container: docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
See 'docker run --help'.
: exit status 125
Traceback (most recent call last):
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connectors/docker.py", line 136, in start
    port = int(self._proc.stdout.readline())
ValueError: invalid literal for int() with base 10: ''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kgb33/Code/dagger/sdk/python/examples/client-simple/say.py", line 32, in <module>
    anyio.run(main, sys.argv[1:])
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_core/_eventloop.py", line 70, in run
    return asynclib.run(func, *args, **backend_options)
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 292, in run
    return native_run(wrapper(), debug=debug)
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 287, in wrapper
    return await func(*args)
  File "/home/kgb33/Code/dagger/sdk/python/examples/client-simple/say.py", line 11, in main
    async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connection.py", line 39, in __aenter__
    return await self.connector.connect()
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connectors/docker.py", line 186, in connect
    await anyio.to_thread.run_sync(self.provision_sync)
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/home/kgb33/Code/dagger/sdk/python/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connectors/docker.py", line 202, in provision_sync
    self.engine.start()
  File "/home/kgb33/Code/dagger/sdk/python/src/dagger/connectors/docker.py", line 146, in start
    raise DaggerException(
dagger.exceptions.DaggerException: Dagger engine failed to start, is docker running?
```

Comment (helderco) [reply sdk/python/src/dagger/connectors/docker.py]: Thank you for the heads up @sipsma, and thanks @KGB33 for quickly addressing this!

I'm planning on doing some cool stuff with `log_output` and IO streams soon so I can deal with it then:
- #3838

File: sdk/python/src/dagger/connectors/docker.py [hunk 2]
```diff
@@ -124,15 +127,30 @@ def start(self) -> None:
...
 
-        # read port number from first line of stdout
-        port = int(self._proc.stdout.readline())
+        try:
+            # read port number from first line of stdout
+            port = int(self._proc.stdout.readline())
+        except ValueError as e:
+            # Check if the subprocess exited with an error.
+            exit_code = self._proc.poll()
+            if exit_code != 0 or exit_code is not None:
```
Comment (helderco) [inline sdk/python/src/dagger/connectors/docker.py:140]: Same as `if exit_code:`.

File: sdk/python/src/dagger/connectors/docker.py [hunk 3]
```diff
@@ -124,15 +127,30 @@ def start(self) -> None:
...
+                    raise DaggerException(
+                        f"Dagger engine failed to start. {self._proc.stderr.readline()}"
+                    ) from e
+                else:
+                    raise DaggerException(
+                        "Dagger engine failed to start, is docker running?"
+                    )
+            # if the subprocess is OK raise the original error
+            else:
+                raise e
```
Comment (helderco) [inline sdk/python/src/dagger/connectors/docker.py:151]: Nit: better at the top and avoid the `else`.

**Review by helderco — APPROVED** [review]
Thanks for working on this @KGB33, it's an improvement for sure!

It needs tests but I can deal with it in #3811.

File: sdk/python/src/dagger/connectors/docker.py (hunk 1 above)
Comment (helderco) [reply sdk/python/src/dagger/connectors/docker.py]: Thank you for the heads up @sipsma, and thanks @KGB33 for quickly addressing this!

I'm planning on doing some cool stuff with `log_output` and IO streams soon so I can deal with it then:
- #3838

**Comment by KGB33** [comment]
Thanks for cleaning it up for me!

I'd be happy to take a stab at some tests over the next few days (either in this PR or anther).

File: sdk/python/src/dagger/connectors/docker.py [hunk 4]
```diff
@@ -91,12 +92,14 @@ def start(self) -> None:
                         encoding="utf-8",
                         check=True,
                     )
+                except FileNotFoundError as e:
+                    raise ProvisionError(
+                        f"Command '{docker_run_args[0]}' not found."
+                    ) from e
                 except subprocess.CalledProcessError as e:
-                    tmp_bin.close()
-                    os.unlink(tmp_bin.name)
```
[the remark this answers is not in the document]
Comment (KGB33) [reply sdk/python/src/dagger/connectors/docker.py:96]: If these two lines need to be kept maybe the pattern described [here](https://stackoverflow.com/a/24174167/10587086) could be useful.

tldr: create a cleanup function & set it to a no-op if there are no errors.
```python
    ...
    def cleanup:
        tmp_bin.close()
        os.unlink(tmp_bin.name)
    try:
        subprocess.run(...)
    except FileNotFoundError as e:
        raise ProvisionError(...) from e
    except subprocess.CalledProcessError as e:
        raise ProvisionError(...) from e
    else: # No errors
        cleanup = lambda: None
    finally:
        cleanup()
    ...
```

[the remark this answers is not in the document]
Comment (helderco) [reply sdk/python/src/dagger/connectors/docker.py:96]: Yeah, love it 🙂

**Comment by helderco** [comment]
Thanks @KGB33, I'm going to merge. Feel free to add tests in a follow up PR 🙂
